Sources/Datadog/Core/Upload/DataUploader.swift

```diff
@@ -6,132 +6,91 @@
-            return urlWithClientToken
-        }
-        return url
+        self.url = urlComponents?.url ?? url
     }
+}
 
-    init(urlWithClientToken: URL, queryItemProviders: [QueryItemProvider]) {
-        self.urlWithClientToken = urlWithClientToken
-        self.queryItemProviders = queryItemProviders
-    }
+/// A type that performs data uploads.
+internal protocol DataUploaderType {
+    func upload(data: Data) -> DataUploadStatus
 }
 
 /// Synchronously uploads data to server using `HTTPClient`.
-internal final class DataUploader {
-    private let urlProvider: UploadURLProvider
+internal final class DataUploader: DataUploaderType {
+    /// An unreachable upload status - only meant to satisfy the compiler.
+    private static let unreachableUploadStatus = DataUploadStatus(needsRetry: false, userDebugDescription: "", userErrorMessage: nil, internalMonitoringError: nil)
+
     private let httpClient: HTTPClient
-    private let httpHeaders: HTTPHeaders
-    private let internalMonitor: InternalMonitor?
+    private let uploadURL: UploadURL
+    private let headersProvider: HTTPHeadersProvider
 
     init(
-        urlProvider: UploadURLProvider,
         httpClient: HTTPClient,
-        httpHeaders: HTTPHeaders,
-        internalMonitor: InternalMonitor? = nil
+        uploadURL: UploadURL,
+        headersProvider: HTTPHeadersProvider
     ) {
-        self.urlProvider = urlProvider
         self.httpClient = httpClient
-        self.httpHeaders = httpHeaders
-        self.internalMonitor = internalMonitor
+        self.uploadURL = uploadURL
+        self.headersProvider = headersProvider
     }
 
-    /// Uploads data synchronously (will block current thread) and returns upload status.
+    /// Uploads data synchronously (will block current thread) and returns the upload status.
     /// Uses timeout configured for `HTTPClient`.
     func upload(data: Data) -> DataUploadStatus {
-        let request = createRequestWith(data: data)
+        let (request, ddRequestID) = createRequestWith(data: data)
         var uploadStatus: DataUploadStatus?
 
         let semaphore = DispatchSemaphore(value: 0)
 
-        httpClient.send(request: request) { [weak self] result in
+        httpClient.send(request: request) { result in
             switch result {
             case .success(let httpResponse):
-                uploadStatus = DataUploadStatus(from: httpResponse)
+                uploadStatus = DataUploadStatus(httpResponse: httpResponse, ddRequestID: ddRequestID)
             case .failure(let error):
-                self?.internalMonitor?.sdkLogger.error("Failed to upload data", error: error)
-                uploadStatus = .networkError
+                uploadStatus = DataUploadStatus(networkError: error)
             }
 
             semaphore.signal()
         }
 
         _ = semaphore.wait(timeout: .distantFuture)
 
-        return uploadStatus ?? .unknown
+        return uploadStatus ?? DataUploader.unreachableUploadStatus
```

Comment:
The `?? DataUploader.unreachableUploadStatus` here is to only satisfy the compiler - the right side of `??` operator will never be evaluated as enforced by the previous logic in this function. That's why the name and comment:
```swift
/// An unreachable upload status - only meant to satisfy the compiler.
private static let unreachableUploadStatus = DataUploadStatus(...)
```
Alternative to this would be:
```swift
guard let uploadStatus = uploadStatus else {
   fatalError()
}
return uploadStatus
```
but as we avoid using `fatalError()` in the SDK codebase, I adopted the solution which is already used in other places: returning known default as fallback.

Hence I don't think we should improve the error value here - this would give a false sense of reachability / usability.